### ISSUE TITLE
Blender 3.0 Support

### ISSUE THREAD

**Issue by rlguy** [rlguy/Blender-FLIP-Fluids#535]
Blender 3.0 is in early development and is set to release on August 25, 2021. At the time of this writing, we have not extensively tested Blender 3.0 and would consider this version not yet officially supported with the FLIP Fluids addon.

Official support will be added at a later date. Typically, the compatibility changes between Blender versions after Blender 2.80 are very small - if there even are any changes needed.

At the moment, there are no known issues when using Blender 3.0 with the FLIP Fluids addon version 1.0.9b or the latest experimental version. The [Blender 3.0 Cycles-X branch](https://builder.blender.org/download/experimental/) also seems to work after some brief testing.

This issue thread will be used to track issues related to the FLIP Fluids addon and Blender Version 3.0.

**Comment by rlguy** [comment]
The latest experimental version of the FLIP Fluids addon (v9.0.9.23) no longer functions in the latest Blender 3.0 Alpha build (July 22, 2021, 64bb49fa4e3c) or in the latest Cycles-X build.

These issues will need to be fixed before the Blender 3.0 release in August 2021 (exact date TBA).

## Known Issues

### Error when setting an object as a domain

This error halts the domain initialization process, which causes the _fluid_surface_ Blender mesh object not to be generated.

```
Traceback (most recent call last):
  File "C:\Users\ryanl\AppData\Roaming\Blender Foundation\Blender\3.0\scripts\addons\flip_fluids_addon\materials\material_library.py", line 30, in get_surface_material_enums_ui
    enums += __get_non_material_library_enums_by_type()
  File "C:\Users\ryanl\AppData\Roaming\Blender Foundation\Blender\3.0\scripts\addons\flip_fluids_addon\materials\material_library.py", line 88, in __get_non_material_library_enums_by_type
    e = (m.name, m.name, "", m.preview.icon_id, __get_non_material_library_material_hash(m))
AttributeError: 'NoneType' object has no attribute 'icon_id'
File "C:\Users\ryanl\AppData\Roaming\Blender Foundation\Blender\3.0\scripts\addons\flip_fluids_addon\materials\material_library.py", line 25, in get_surface_material_enums_ui
Traceback (most recent call last):
  File "C:\Users\ryanl\AppData\Roaming\Blender Foundation\Blender\3.0\scripts\addons\flip_fluids_addon\materials\material_library.py", line 30, in get_surface_material_enums_ui
    enums += __get_non_material_library_enums_by_type()
  File "C:\Users\ryanl\AppData\Roaming\Blender Foundation\Blender\3.0\scripts\addons\flip_fluids_addon\materials\material_library.py", line 88, in __get_non_material_library_enums_by_type
    e = (m.name, m.name, "", m.preview.icon_id, __get_non_material_library_material_hash(m))
AttributeError: 'NoneType' object has no attribute 'icon_id'
File "C:\Users\ryanl\AppData\Roaming\Blender Foundation\Blender\3.0\scripts\addons\flip_fluids_addon\materials\material_library.py", line 25, in get_surface_material_enums_ui
Traceback (most recent call last):
  File "C:\Users\ryanl\AppData\Roaming\Blender Foundation\Blender\3.0\scripts\addons\flip_fluids_addon\properties\object_properties.py", line 102, in <lambda>
    set=lambda self, value: self._set_object_type(value),
  File "C:\Users\ryanl\AppData\Roaming\Blender Foundation\Blender\3.0\scripts\addons\flip_fluids_addon\properties\object_properties.py", line 245, in _set_object_type
    active_object.flip_fluid.domain.initialize()
  File "C:\Users\ryanl\AppData\Roaming\Blender Foundation\Blender\3.0\scripts\addons\flip_fluids_addon\properties\domain_properties.py", line 148, in initialize
    self.materials.initialize()
  File "C:\Users\ryanl\AppData\Roaming\Blender Foundation\Blender\3.0\scripts\addons\flip_fluids_addon\properties\domain_materials_properties.py", line 98, in initialize
    self.surface_material = 'MATERIAL_NONE'
TypeError: bpy_struct: item.attr = val: enum "MATERIAL_NONE" not found in ()
```

### Error when trying to bake a simulation

Generated upon pressing the _Bake_ operator. This error halts the baking process and prevents the simulator from starting.

```
Error: Python: Traceback (most recent call last):
  File "C:\Users\ryanl\AppData\Roaming\Blender Foundation\Blender\3.0\scripts\addons\flip_fluids_addon\operators\export_operators.py", line 148, in modal
    self._export_simulation_data_file()
  File "C:\Users\ryanl\AppData\Roaming\Blender Foundation\Blender\3.0\scripts\addons\flip_fluids_addon\operators\export_operators.py", line 101, in _export_simulation_data_file
    dprops.bake.export_success = export.export_simulation_data(
  File "C:\Users\ryanl\AppData\Roaming\Blender Foundation\Blender\3.0\scripts\addons\flip_fluids_addon\export.py", line 431, in export_simulation_data
    __export_simulation_data_to_file(context, simulation_objects, data_filepath)
  File "C:\Users\ryanl\AppData\Roaming\Blender Foundation\Blender\3.0\scripts\addons\flip_fluids_addon\export.py", line 38, in __export_simulation_data_to_file
    data = __get_simulation_data_dict(context, simobjects)
  File "C:\Users\ryanl\AppData\Roaming\Blender Foundation\Blender\3.0\scripts\addons\flip_fluids_addon\export.py", line 48, in __get_simulation_data_dict
    data['domain_data'] = __get_domain_data_dict(context, simobjects.domain)
  File "C:\Users\ryanl\AppData\Roaming\Blender Foundation\Blender\3.0\scripts\addons\flip_fluids_addon\export.py", line 109, in __get_domain_data_dict
    d['advanced']['num_threads_auto_detect'] = dprops.advanced.num_threads_auto_detect
KeyError: 'advanced'
```

### Immediate render crash

If attempting to render an already baked cache (animation or frame), Blender will immediately crash with an _EXCEPTION ACCESS VIOLATION_ error. This crash seems to occur during frame load before rendering begins.

**Comment by rlguy** [comment]
Fixes for the above issues are now available in [FLIP Fluids Addon Experimental 9.0.9.24](https://github.com/rlguy/Blender-FLIP-Fluids/wiki/Experimental-Builds). These fixes will also be available in the next stable release next week, FLIP Fluids v1.1.0.

At the moment there are no known compatibility issues with this version and the daily Blender 3.0 Alpha release.

**Comment by rlguy** [comment]
Due to recent API changes in Blender 3.0, two minor issues had arisen:

- The documentation link in the FLIP Fluids Preferences Menu had disappeared.
- The automatic Cycles visibility toggles upon creating FLIP Fluid object had stopped functioning.

These have been fixed and are now available in the latest experimental build as of version 9.1.1.

**Comment by rlguy** [comment]
FLIP Fluids Experimental Version 9.1.2 (03-NOV-2021) adds compatibility updates for recent versions of Blender 3.0 Beta.

From [Release Notes](https://github.com/rlguy/Blender-FLIP-Fluids/wiki/Release-Notes#experimental-versions-91x):

- Blender 3.0 Compatibility: This build is intended to be used with Blender 3.0 Beta builds dated November 3, 2021 or later.
    - Added: Forward compatibility support for Blender 3.0
        - If a .blend file is saved in Blender 3.0 and then opened in an earlier version of Blender 2.8x or 2.9x, the addon will automatically update the .blend file so that FLIP Fluid objects are compatible and function correctly.
        - We still recommend keeping Blender 3.0 .blend files and earlier versions in separate saves. Blender will warn that you may still experience data loss in the .blend file in areas/features outside of the FLIP Fluids addon.
    - Bug Fix: Issues related to automatically setting Cycles ray visibility for FLIP Fluid/Inflow/Outflow/Force/Domain objects in recent builds of Blender 3.0.

**Comment by rlguy** [comment]
The latest stable release, FLIP Fluids 1.2.0, adds official support for Blender 3.0.

Blender 3.0 releases on December 3rd, 2021 and at this point any changes before release should not break compatibility.

**Comment by SebastienLacombre** [comment]
La version 1.2.0 de Flip Fluid ne fonctionne pas avec Blender 3.0 sous Ma
![System](https://user-images.githubusercontent.com/60298826/144318424-6cebc787-61c4-4c6d-9a11-964293699c2e.jpg)
![VersionBlender](https://user-images.githubusercontent.com/60298826/144318439-3e853a6e-4085-4e13-8be1-ddeb9e89f9ec.jpeg)
![Version FlipFluid](https://user-images.githubusercontent.com/60298826/144318453-1ec0a614-f568-499d-be24-3e8b460cec1c.jpeg)
c
![erreur FlipFluid](https://user-images.githubusercontent.com/60298826/144318464-fc7d2a48-57a9-4ae4-ac61-66d0bca7f4e3.jpeg)

**Comment by rlguy** [comment]
Thanks for the report! We recently became aware of this issue and are working on a fix. This fix should become available before the Blender 3.0 release (December, 3, 2021) or early next week at the latest. I will update this thread when a solution is available.

This issue may affect artists that are running macOS versions that are [no longer supported by Apple](https://endoflife.date/macos) and are not receiving service updates. This does not affect macOS 10.14 (Mojave), but affects versions 10.13 (High Sierra) and lower.

A current workaround is to update the macOS version, but this may not be a solution that is possible or desired by some artists.

---

Merci pour le rapport ! Nous avons récemment pris connaissance de ce problème et travaillons sur un correctif. Ce correctif devrait être disponible avant la sortie de Blender 3.0 (3 décembre 2021) ou au plus tard au début de la semaine prochaine. Je mettrai à jour ce fil dès qu'une solution sera disponible.

Ce problème peut affecter certains artistes qui exécutent des versions de macOS qui ne sont [plus prises en charge par Apple](https://endoflife.date/macos) et ne reçoivent pas de mises à jour de service. Cela n'affecte pas macOS 10.14 (Mojave), mais affecte les versions 10.13 (High Sierra) et inférieures.

**Comment by SebastienLacombre** [comment]
Merci pour votre réponse. Je comprends mieux 😁 C'est pas grave. Je vais
continuer avec la version 1.0.9b de Flipfluid sur la version 2.93. Celle-ci
fonctionne très bien.

**Comment by rlguy** [comment]
Hi, we have just added a fix for this in FLIP Fluids Version 1.2.1. This new version can be accessed in the Blender Market downloads with this installer: _FLIP_Fluids_addon_1.2.1_stable_(03_dec_2021).zip_.

If you test this version and are still experiencing issues, please let us know, as this could indicate a different issue.

---

Bonjour, nous venons d'ajouter un correctif pour cela dans FLIP Fluids Version 1.2.1. Cette nouvelle version est accessible dans les téléchargements de Blender Market avec cet installateur : _FLIP_Fluids_addon_1.2.1_stable_(03_dec_2021).zip_.

Si vous testez cette version et que vous rencontrez toujours des problèmes, veuillez nous en informer, car cela pourrait indiquer un problème différent.

**Comment by SebastienLacombre** [comment]
Oooooh ! Vous êtes le meilleur ! Ça fonctionne parfaitement ! Merci, merci, merci beaucoup ! C’est un vrai bonheur :D
